openml/study/functions.py

```diff
@@ -56,44 +91,81 @@ def get_study(study_id, entity_type=None):
-        flows = [int(x) for x in result_dict['oml:flows']['oml:flow_id']]
-    if 'oml:setups' in result_dict:
-        setups = [int(x) for x in result_dict['oml:setups']['oml:setup_id']]
-    if 'oml:runs' in result_dict:
-        runs = [int(x) for x in result_dict['oml:runs']['oml:run_id']]
-
-    study = OpenMLStudy(
-        study_id=study_id,
-        alias=alias,
-        main_entity_type=main_entity_type,
-        benchmark_suite=benchmark_suite,
-        name=name,
-        description=description,
-        status=status,
-        creation_date=creation_date,
-        creator=creator,
-        tags=tags,
-        data=datasets,
-        tasks=tasks,
-        flows=flows,
-        setups=setups,
-        runs=runs
-    )
+    else:
+        raise ValueError('No tasks attached to study {}!'.format(id_))
+
+    if main_entity_type in ['runs', 'run']:
+
+        if 'oml:flows' in result_dict:
+            flows = [int(x) for x in result_dict['oml:flows']['oml:flow_id']]
+        else:
+            raise ValueError('No flows attached to study {}!'.format(id_))
+        if 'oml:setups' in result_dict:
+            setups = [int(x) for x in result_dict['oml:setups']['oml:setup_id']]
+        else:
+            raise ValueError('No setups attached to study!'.format(id_))
+        if 'oml:runs' in result_dict:
+            runs = [
+                int(x) for x in result_dict['oml:runs']['oml:run_id']
+            ]  # type: Optional[List[int]]
+        else:
+            if creation_date_as_date < dateutil.parser.parse('2019-01-01'):
+                # Legacy studies did not require runs
+                runs = None
+            else:
+                raise ValueError('No runs attached to study!'.format(id_))
+
+        study = OpenMLStudy(
+            study_id=study_id,
+            alias=alias,
+            benchmark_suite=benchmark_suite,
+            name=name,
+            description=description,
+            status=status,
+            creation_date=creation_date,
+            creator=creator,
+            tags=tags,
+            data=datasets,
+            tasks=tasks,
+            flows=flows,
+            setups=setups,
+            runs=runs,
+        )  # type: BaseStudy
+
+    elif main_entity_type in ['tasks', 'task']:
+
+        study = OpenMLBenchmarkSuite(
+            suite_id=study_id,
+            alias=alias,
+            name=name,
+            description=description,
+            status=status,
+            creation_date=creation_date,
+            creator=creator,
+            tags=tags,
+            data=datasets,
+            tasks=tasks
+        )
+
+    else:
+        raise ValueError('Unknown entity type {}'.format(main_entity_type))
+
     return study
 
 
-def create_study(alias, benchmark_suite, name, description, run_ids):
+def create_study(
+    alias: Optional[str],
+    benchmark_suite: Optional[int],
+    name: str,
+    description: str,
+    run_ids: List[int],
+) -> OpenMLStudy:
```

Comment:
Can we put non-optional before optional?
If parameters are optional, they need a default value, otherwise they are enforced.